### PR TITLE
api, feeds: wire-in feeds resolution

### PR DESCRIPTION
**Pull request by acud**
This PR integrates the feeds package, providing API endpoints to lookup (and update) feeds.
Currently calling the bee  signer is not supported, so the update  endpoint needs  to  supply the signature.
However, for the signature, both the next update index and the wrapped chunk('s hash) are  needed.
so it can only be used after a lookup. 
The lookup endpoint returns a  response with header values for `swarm-feed-index` and `swarm-feed-index-next` specifying the 32-byte current and next index in hex encoding.

NOTE: this feature  is not implemented for epochs only sequence-indexing.

The PR also provides support for automatic feed resolution on the `bzz` endpoint.
This requires the feed attributes to  be encoded as entry metadata in a mantaray entry  on the '/' path.
This is a temporary workaround until mantaray is fixedd to support  entry on the empty path.
such a manifest is created by calling  the `feeds` POST endpoint

The API is  detailed in the opanapi spec

### CODE REVIEW

File: openapi/Swarm.yaml [hunk 1]
```diff
@@ -818,3 +818,162 @@ paths:
...
+            $ref: 'SwarmCommon.yaml#/components/schemas/HexString'
+          required: false
+          description: Feed update index
+        - in: query
+          name: type
+          schema:
+            $ref: 'SwarmCommon.yaml#/components/schemas/FeedType'
+          required: false
+          description: Feed indexing scheme (default: sequence)
+        - in: query
```
Comment (agazso) [inline openapi/Swarm.yaml:889]: Why do we need the signature for finding a feed update?

File: openapi/Swarm.yaml [hunk 2]
```diff
@@ -818,3 +818,162 @@ paths:
...
+                $ref: 'SwarmCommon.yaml#/components/schemas/ReferenceResponse'
+        '400':
+           $ref: 'SwarmCommon.yaml#/components/responses/400'
+        '401':
+           $ref: 'SwarmCommon.yaml#/components/responses/401'
+        '500':
+           $ref: 'SwarmCommon.yaml#/components/responses/500'
+        default:
+          description: Default response
+    get:
```
Comment (agazso) [inline openapi/Swarm.yaml:860]: The time (`at`) seems to be missing here if I'm understand correctly.

File: openapi/Swarm.yaml (hunk 1 above)
Comment (zelig) [reply openapi/Swarm.yaml:889]: this is messed up  yes, post params are under get.

`at` should be param in all 3 methods

File: openapi/Swarm.yaml [hunk 3]
```diff
@@ -818,3 +818,162 @@ paths:
...
+          name: type
+          schema:
+            $ref: 'SwarmCommon.yaml#/components/schemas/FeedType'
+          required: false
+          description: Feed indexing scheme (default: sequence)
+        - in: query
+          name: sig
+          schema:
+            $ref: 'SwarmCommon.yaml#/components/schemas/HexString'
+          required: false
```
Comment (agazso) [inline openapi/Swarm.yaml:962]: I think the signature should be required

Comment (acud) [reply openapi/Swarm.yaml:962]: :+1: setting it to `true`. once we implement fallback to internal signer this can be switched off

Comment (agazso) [reply openapi/Swarm.yaml]: Isn't it a security issue if the users of this endpoint can pass in any data and Bee signs it? Or is this functionality is going to be private somehow to trusted users?

File: pkg/api/feed.go [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+			return
+		}
+	}
+
+	var at int64
+	atStr := r.URL.Query().Get("at")
+	if atStr != "" {
+		at, err = strconv.ParseInt(atStr, 10, 64)
+		if err != nil {
+			panic(err)
```
Comment (zelig) [inline pkg/api/feed.go:57]: panics will be cleaned up yeah?

File: pkg/api/feed.go [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+	ch, err := lookup.At(r.Context(), int64(at), 0)
+	if err != nil {
+		panic(err)
+	}
+
+	update := ch.Data()
+
+	// split the timestamp and reference
+	if len(update) < swarm.SpanSize*2 {
+		panic("malformed data")
```
Comment (zelig) [inline pkg/api/feed.go:77]: why not explicitly require 8+8+32=48 or 8+8+64=80  bytes

File: pkg/api/feed_test.go [hunk 6]
```diff
@@ -0,0 +1,383 @@
...
+
+		if h := respHeaders[api.SwarmFeedIndexHeader]; len(h) > 0 {
+			b, err := hex.DecodeString(h[0])
+			if err != nil {
+				t.Fatal(err)
+			}
+			if !bytes.Equal(b, timestampBytes) {
+				t.Fatalf("feed index header mismatch. got %v want %v", b, timestampBytes)
+			}
+		} else {
```
Comment (zelig) [inline pkg/api/feed_test.go:128]: check if len(header) == 0  then no need for else

File: pkg/api/feed_test.go [hunk 7]
```diff
@@ -0,0 +1,383 @@
...
+		return l.chunk, nil
+	}
+	fmt.Println(at, after, l.at, l.after)
+	if at == l.at && after == l.after {
+		return l.chunk, nil
+	}
+	return nil, errors.New("no feed update found")
+}
+
+func mockUpdate(t int64, a swarm.Address) swarm.Chunk {
```
Comment (zelig) [inline pkg/api/feed_test.go:339]: no need to have a mock when the mock does what the real thing does. just use the feed toChunk method etc

File: pkg/api/feed_test.go [hunk 8]
```diff
@@ -0,0 +1,383 @@
...
+	}
+	_, err = hasher.Write(b)
+	if err != nil {
+		panic(err)
+	}
+	s := swarm.NewAddress(hasher.Sum(nil))
+	return swarm.NewChunk(s, append(b, tb...))
+}
+
+func socChunk(t *testing.T, ch swarm.Chunk) (*soc.Soc, []byte, []byte) {
```
Comment (zelig) [inline pkg/api/feed_test.go:362]: here too. lets not duplicate code

File: pkg/api/feed.go (hunk 4 above)
Comment (acud) [reply pkg/api/feed.go:57]: of course

File: pkg/feeds/epochs/finder.go [hunk 9]
```diff
@@ -164,10 +165,14 @@ func (f *asyncFinder) at(ctx context.Context, at int64, p *path, e *epoch, c cha
 		}
 	}
 }
+func (f *asyncFinder) At(ctx context.Context, at, after int64) (swarm.Chunk, feeds.Index, feeds.Index, error) {
```
Comment (zelig) [inline pkg/feeds/epochs/finder.go:168]: comment FIXME/TODO that this needs to be implemented

File: pkg/feeds/factory/factory.go [hunk 10]
```diff
@@ -0,0 +1,31 @@
+// Copyright 2021 The Swarm Authors. All rights reserved.
```
Comment (zelig) [inline pkg/feeds/factory/factory.go:1]: TODO: This needs to be implemented directly on the feed struct which should have an index type field.

File: pkg/feeds/getter.go [hunk 11]
```diff
@@ -17,7 +17,7 @@ import (
 
 // Lookup is the interface for time based feed lookup
 type Lookup interface {
-	At(ctx context.Context, at, after int64) (swarm.Chunk, error)
+	At(ctx context.Context, at, after int64) (chunk swarm.Chunk, currentIndex, nextIndex Index, err error)
```
Comment (zelig) [inline pkg/feeds/getter.go:20]: TODO: in fact At when used with noncurrent time should not return a next index.
only  latest  should..

**Review by zelig — APPROVED** [review]

File: pkg/api/bzz.go [hunk 12]
```diff
@@ -69,6 +78,34 @@ func (s *server) bzzDownloadHandler(w http.ResponseWriter, r *http.Request) {
...
+
+	// there's a possible ambiguity here, right now the data which was
+	// read can be an entry.Entry or a mantaray feed manifest. Try to
+	// unmarshal as mantaray first and possibly resolve the feed, otherwise
+	// go on normally.
+	if !feedDereferenced {
+		if l, err := s.manifestFeed(ctx, ls, buf.Bytes()); err == nil {
+			//we have a feed manifest here
+			ch, cur, next, err := l.At(ctx, time.Now().Unix(), 0)
+			fmt.Println(cur, next)
```
[the remark this answers is not in the document]
Comment (acud) [reply pkg/api/bzz.go:90]: :+1: forgot about this. added the current index in the return headers. this was missing :+1:

File: pkg/feeds/factory/factory.go (hunk 10 above)
Comment (acud) [reply pkg/feeds/factory/factory.go:1]: but then how do you inject the mock factory with the mock update?

File: openapi/Swarm.yaml [hunk 13]
```diff
@@ -861,11 +861,107 @@ paths:
...
+          $ref: 'SwarmCommon.yaml#/components/responses/400'
         '401':
-           $ref: 'SwarmCommon.yaml#/components/responses/401'
+          $ref: 'SwarmCommon.yaml#/components/responses/401'
         '500':
            $ref: 'SwarmCommon.yaml#/components/responses/500'
         default:
           description: Default response
 
+  '/feed/{owner}/{topic}':
```
Comment (agazso) [inline openapi/Swarm.yaml:872]: Signature seems to be missing.

Comment (acud) [reply openapi/Swarm.yaml:872]: not needed for `POST` since no SOC is being created

Comment (agazso) [inline openapi/Swarm.yaml:872]: In `router.go` the endpoint is called `/feeds`

https://github.com/ethersphere/bee/blob/aa61c6532d8b7c57c5776250c62264e333feef35/pkg/api/router.go#L92

**Review by zelig — APPROVED** [review]

**Pull request merged**